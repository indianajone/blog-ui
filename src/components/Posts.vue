<template>
  <div class="posts level is-verticle">
    <router-link class="button ml-auto is-primary is-outlined" :to="'/create'">
        New Post
    </router-link>
    <table class="table is-fullwidth is-striped">
      <thead>
        <tr>
          <th class="post-title" @click="sortBy('title')">Title</th>
          <th class="post-date is-small" @click="sortBy('createdAt')">Date</th>
        </tr>
      </thead>
      <tbody>
        <tr class="post-item" v-for="post in sortedPosts" :key="post._id">
          <td>
             <div class="media" >
              <figure class="media-left" v-if="post.image">
                <p class="image is-32x32">
                  <img :src="getImage(post.image)">
                </p>
              </figure>
              <div class="media-content">
                <div class="content">
                  <p class="level">
                    <router-link :to="'/post/' + post._id">
                      {{ post.title }}
                    </router-link>
                  </p>
                </div>
              </div>
            </div>
          </td>
          <td>{{ post.createdAt | fromNow }}</td>
        </tr>
      </tbody>
    </table>
  </div>
</template>

<script>
import * as _ from 'lodash';

export default {
    name: 'Posts',
    data() {
        return {
            posts: null,
            key: 'createdAt',
            reverse: true
        };
    },
    mounted() {
        this.fetch();
    },
    computed: {
        sortedPosts() {
            return _.orderBy(this.posts, [this.key], [this.reverse ? 'desc' : 'asc']);
        }
    },
    methods: {
        getImage(path) {
            return `${this.$http.defaults.baseURL}/${path}`;
        },
        sortBy(key) {
            this.key = key;
            this.reverse = !this.reverse;
        },
        async fetch() {
            try {
                const { data } = await this.$http.get('/posts');
                this.posts = data.data;
            } catch (e) {
                console.warn('Cannot connect to the server.');
            }
        }
    }
};
</script>

<style scoped lang="scss">
@import "~bulma/sass/utilities/_all";
@import "~bulma/sass/layout/section";
.posts {
    @extend .section;
    th {
        cursor: pointer;
    }
    .is-small {
        width: 30%;
    }
}
</style>
