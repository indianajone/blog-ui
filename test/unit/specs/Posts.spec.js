import { mount } from 'vue-test-utils';
import Posts from '@/components/Posts';
import { createRenderer } from 'vue-server-renderer';

describe('Posts.vue', () => {

    let wrapper;

    beforeEach(() => {
        wrapper = mount(Posts, {
            mocks: {
                $http: {
                    defaults: { baseURL: 'localhost' },
                    get() {
                        return { data: {
                            data: [
                                {
                                    title: 'New Post',
                                    body: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit.',
                                    _id: '59eccf0e4e8b8f72705ac87d',
                                    createdAt: '2017-10-02T17:02:06.051Z',
                                    type: 'text'
                                },
                                {
                                    title: 'Other Post',
                                    body: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit.',
                                    _id: '59eccf0e4e8b8f72705ac87d',
                                    createdAt: '2017-11-22T17:01:16.051Z',
                                    type: 'text'
                                },
                                {
                                    title: 'Another Post',
                                    body: 'Lorem ipsum dolor sit amet, consectetur adipisicing elit.',
                                    _id: '59eccf0e4e8b8f72705ac87d',
                                    createdAt: '2016-07-18T17:01:16.051Z',
                                    type: 'text'
                                }
                            ]
                        } };
                    }
                }
            }
        });
    });

    it('should render all posts.', () => {
        const rows = wrapper.findAll('.post-item');

        expect(rows.at(0).find('p').text()).toContain('Other Post');
        expect(rows).toHaveLength(3);
    });

    it('should return full image path.', () => {
        expect(wrapper.vm.getImage('/uploads/image.jpg'))
            .toEqual('localhost//uploads/image.jpg');
    });

    it('should sort posts by alphabet.', () => {
        const button = wrapper.find('.post-title');
        button.trigger('click');

        expect(wrapper.findAll('.post-item').at(0).find('p').text()).toContain('Another Post');
    });

    it('should matches snapshot', () => {
        const renderer = createRenderer();
        renderer.renderToString(wrapper.vm, (err, str) => {

            if (err) {
                throw new Error(err);
            }
            expect(str).toMatchSnapshot();

        });

    });

});
