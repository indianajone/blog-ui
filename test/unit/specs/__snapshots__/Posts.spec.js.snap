// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Posts.vue should matches snapshot 1`] = `
<div class="posts level is-verticle">
    <router-link to="/create" class="button ml-auto is-primary is-outlined">
        New Post
    </router-link>
    <table class="table is-fullwidth is-striped">
        <thead>
            <tr>
                <th class="post-title">Title</th>
                <th class="post-date is-small">Date</th>
            </tr>
        </thead>
        <tbody>
            <tr class="post-item">
                <td>
                    <div class="media">
                        <!---->
                        <div class="media-content">
                            <div class="content">
                                <p class="level">
                                    <router-link to="/post/59eccf0e4e8b8f72705ac87d">
                                        Other Post
                                    </router-link>
                                </p>
                            </div>
                        </div>
                    </div>
                </td>
                <td>2017-11-22T17:01:16.051Z</td>
            </tr>
            <tr class="post-item">
                <td>
                    <div class="media">
                        <!---->
                        <div class="media-content">
                            <div class="content">
                                <p class="level">
                                    <router-link to="/post/59eccf0e4e8b8f72705ac87d">
                                        New Post
                                    </router-link>
                                </p>
                            </div>
                        </div>
                    </div>
                </td>
                <td>2017-10-02T17:02:06.051Z</td>
            </tr>
            <tr class="post-item">
                <td>
                    <div class="media">
                        <!---->
                        <div class="media-content">
                            <div class="content">
                                <p class="level">
                                    <router-link to="/post/59eccf0e4e8b8f72705ac87d">
                                        Another Post
                                    </router-link>
                                </p>
                            </div>
                        </div>
                    </div>
                </td>
                <td>2016-07-18T17:01:16.051Z</td>
            </tr>
        </tbody>
    </table>
</div>
`;
